### DISCUSSION TITLE
Destroy/Clean up a machine stopped in an arbitrary moment (eg, React DOM unmount)

### DISCUSSION THREAD

**Discussion by VanTanev** [statelyai/xstate#1825] [Q&A] [answered]
I have a machine that can enter a state that invokes a service which needs to be cleaned up, and I'm having a hard time modelling it.

Outside XState, the service works like this:
```typescript
const tabLock= new TabLock()

// acquire a lock:
tabLock.acquire() // returns a Promise<void> that fulfills on success and rejects on failure

// release a lock
tabLock.release() // synchronously releases the lock

// destroy all resources (eg BroadcastChannel)
tabLock.destroy()
```

```typescript
// in react:
function useTabLock() {
  const [tabLock] = React.useState(() => new TabLock())
  React.useEffect(() => {
    // cleanup tabLock
    () => tabLock.destroy()
  }, [tabLock]
  return tabLock
}
```

In XState, I initially did this:
```typescript
const machineCreator = () => createMachine({
  context: {
    tabLock: new TabLock()
  },
  initial: 'idle',
  states: {
    idle: {
      on: { ACQUIRE: 'acquiring' }
      on: { DESTROY: 'destroy' }
    },
    acquiring: {
      invoke: {
        src: async (ctx) => {
          await ctx.tabLock.aquire() // first acquires a browser-level lock
          await acquireNetworkLock(ctx) // acquires network-level lock
        },
        onDone: 'locked',
        onError: 'idle'
      }
    },
    locked: {
      on: { RELEASE: 'releasing', }
    },
    releasing: {
      invoke: {
        src: async (ctx) => {
            await releaseNetworkLock(ctx) // releases the network-level lock
            await ctx.tabLock.release()
          }
        },
        onDone: 'idle'
      }
    },
    destroyed: {
      type: 'final',
      entry: (ctx) => ctx.tabLock.destroy()
    }
  }
})
```

The above however breaks when the user navigates away without going through a "release." 
The especially problematic part is `tabLock.destroy()` not being called, and the object just hangs around without a way to clean it up.

**Comment by VanTanev** [comment]
**Edit:** There is a better solution: https://github.com/davidkpiano/xstate/discussions/1825#discussioncomment-270799

Original below:

-----

The one solution I found was this:
```typescript
const machineCreator = () => createMachine({
  context: {
    tabLock: new TabLock()
  },
  activities: ['tabLockDestroyer']
  // ...
}, {
  activities: {
    tabLockDestroyer: ctx => () => ctx.tabLock.destroy()
  }
})
```

It feels somewhat bludgeoning, but maybe it's the only way?

**Comment by davidkpiano** [comment]
It only seems hard because you're using `async/await` syntax. This is the same reason that you can't (shouldn't) use `async/await` in `useEffect` - you can't immediately return a cleanup function.

Consider something like this instead:

```js
      invoke: {
        src: (ctx, e) => (sendBack, receive) => {
          new Promise(async () => {
            await someLock.acquire();
          });

          return () => someLock.destroy();
        }
      },
```

**Reply by VanTanev** [reply]
This doesn't work. I probably didn't explain my constraints correctly:

1) This will immediately destroy the lock after acquisition, because the states are `idle -> acquiring -> locked`, and this `invoke` is in `acquiring`. We definitely don't want that, the lock should only be released by the `releasing` state.
2) While the machine is running, a lock might be acquired/released zero, one or many times. Ideally for every `acquire()` there should be a `release()`, but even if there were 0 `acquires`, `destroy()` MUST be called when the machine is stopped to release resources held internally by the `TabLock` object. The design of TabLock could be changed to not require that, but since a machine can be stopped while it's in any arbitrary state, we still want `tabLock.destroy()` to be called on `machine.stop()`. The only way to do that seems to be:  https://github.com/davidkpiano/xstate/discussions/1825#discussioncomment-269603

Tangentially, I really wish XState Activities received a `send()` function. It would make them much more useful, I think?

**Reply by davidkpiano** [reply]
> Tangentially, I really wish XState Activities received a send() function. It would make them much more useful, I think?

This describes invocations, which should be used instead of activities (which are deprecated in v5 - invocations do the same thing).

**Reply by VanTanev** [reply]
@davidkpiano You are correct. And, the answer I posted above can be handled with a machine-wide invoke. I'll update it accordingly.

Back to the original code, the thing that usually makes promise invocations better than callback invocations is the ease of "get a result out of some asynchronous process." With a callback invocation, I would need to specify new event types to handle the functionality of `onDone/onError`, and manually wire those in. 

On the other hand, promise invocations cannot respond to the machine `stop` event. Only callback invocations can.

It's tradeoffs all the way down, I guess 🤷

**Reply by Andarist** [reply]
You could specify a top-level `exit` action that would release the lock - I think this is what you want, right? To free the lock when the machine gets stopped.

**Reply by VanTanev** [reply]
@davidkpiano Actually, converting the `activity` to an `invoke` does not work!

Even if the `invoke` is on the top of the machine, it will be destroyed immediately. I need it to be destroyed when the machine is stopped. For this, it seems like only an activity works.

~It's kinda bad that with v5, it will become impossible to clean up machine-wide resources :-/~ I spoke out of turn.

**Reply by VanTanev** [reply]
@Andarist This is the approach I was looking for, thanks!
I'll rewrite it as a separate answer and mark it as accepted.

**Reply by VanTanev** [reply]
@Andarist @davidkpiano Conceptually, shouldn't a callback invoke at the top level be destroyed only after the machine is stopped? Why does it get destroyed immediately?

**Reply by Andarist** [reply]
Is it maybe just recreated? Maybe the state in which this is declared gets reentered?

**Reply by VanTanev** [reply]
@Andarist It's immediately destroyed, even though the machine is still running: https://codesandbox.io/s/dazzling-hellman-bwqni?file=/src/index.js

~This seems possibly like a bug?~
Wrong!

I was doing the invoke wrong, it's tripple callback, not double, here is the proper version, it works as expected:
https://codesandbox.io/s/friendly-matsumoto-k0mc0?file=/src/index.js

It's kind of easy to get the callback invoke wrong, and TypeScript was no help here :(

**Reply by Andarist** [reply]
The next major version should be more helpful around this in terms of TypeScript. The API will change slightly so it should actually make it clearer how the callback service should be written.

**Comment by VanTanev** [comment, marked as answer]
@Andarist suggested the proper solution [here](https://github.com/davidkpiano/xstate/discussions/1825#discussioncomment-270789):
```typescript
const machineCreator = () => createMachine({
  context: {
    tabLock: new TabLock()
  },
  exit: (ctx) => ctx.tabLock.destroy()
})
```

**Reply by Andarist** [reply]
A more proper one would be to create the `tabLock` in the `entry` action or use a factory function to declare `context`. 

`createMachine` creates "templates" for machines - you can think about it as about creating a *class*. I'm pretty sure you want to keep the lock per instance and not as a static value on the class itself.

**Reply by VanTanev** [reply]
You are, of course, correct. The actual code uses a creator function. I was overly brief in my example code, will update it as to not confuse others.
